packages/mgt-spfx-utils/README.md

```diff
@@ -0,0 +1,157 @@
+# Utilities for SharePoint Framework Web Parts using Microsoft Graph Toolkit
+
+[![npm](https://img.shields.io/npm/v/@microsoft/mgt-spfx-utils?style=for-the-badge)](https://www.npmjs.com/package/@microsoft/mgt-spfx-utils)
+
+![SPFx 1.15.2](https://img.shields.io/badge/SPFx-1.15.2-green.svg?style=for-the-badge)
+
+Helper functions to simplify lazy loading of Microsoft Graph Toolkit components when using disambiguated web components in SharePoint Framework web parts. For more information on disambiguation in MGT see https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/packages/mgt-components#disambiguation
+
+## Installation
+
+To lazy load Microsoft Graph Toolkit components from the library, add the `@microsoft/mgt-spfx-utils` package as a dependency to your SharePoint Framework project. If you use React, also add the `@microsoft/mgt-react` package:
+
+```bash
+npm install @microsoft/mgt-spfx-utils
+```
+
+or
+
+```bash
+yarn add @microsoft/mgt-spfx-utils
+```
+
+or when using React:
+
+```bash
+npm install @microsoft/mgt-spfx-utils @microsoft/mgt-react
+```
+
+or
+
+```bash
+yarn add @microsoft/mgt-spfx-utils @microsoft/mgt-react
+```
+
+## Usage
+
+By disambiguating tag names of Microsoft Graph Toolkit components, you can use your own version of MGT rather than using the centrally deployed `@microsoft/mgt-spfx` package. This allows you to avoid colliding with SharePoint Framework components built by other developers. When disambiguating tag names, MGT is included in the generated SPFx bundle, increasing its size.
```

Comment:
Should we say that mgt-spfx is still valid and supported for 2.x version? But shouldn't be used for > v3?

Reply:
Good point! Also, let's give a recommendation for using MGT v2 and v3 side by side, which is likely to happen given the current adoption and the fact that not everyone will be able to upgrade to v3 right away.

Reply:
I've been thinking about this case a bit lately. I think that we could continue to offer `mgt-spfx` but we'd have to be very clear that you can only have a single `mgt-spfx` version in a tenant, so if folks need to support v2 and v3 in a single tenant then they need to use `mgt-spfx@2` and then disambiguation for their v3 based solutions.

If they only need to support v3 then they could use `mgt-spfx@3` to gain the benefit of smaller solution bundles, at the expense of having all of mgt loaded in the library solution.

Reply:
I'm wondering if we can bring it even another step further. Could we offer a default disambiguation string based on the solution name? So no work for the devs, automatically disambiguated? And you could turn it off or customize it if you really need it?

Reply:
While I like the idea of making it easy for developers, I dislike "magic" black boxes. Plus there's a fair bit of complexity in delivering this.
When should we automatically add a disambiguation? They're only really needed when there might be multiple instances of mgt in a single page, which is __mostly__ in SPFx solutions.
Where are we sourcing the name from? Bearing in mind that wherever your source is only existing in the application importing `@microsoft/mgt-element` (or one of its dependents)
I'd rather we added strong guidance and revisit the magic idea at a later date if there is customer demand.

Reply:
Could this "magic" be part of the mgt-spfx specific packages? It's very much a limitation of web components I'd rather "hide" to our developers, if possible. It's our implementation that cause these side effects...

Reply:
If we were to add this functionality, it would live in `mgt-spfx-utils` IMO.
I'm struggling to see a good way of implementing it in deterministic manner without requiring the developer to pass in an extra input.
As such I strongly suggest that we tackle this auto disambiguation concept as a separate work item.

Reply:
We've got a few ways we could go about it:

- we could use the project name from project.json - in theory it should be unique, because it's used to generate the name for the sppkg file which is deployed to the app catalog
- we could use the project name from .yo-rc.json - my gut feeling tells me it's not a super visible location though, so wouldn't be my preference
- we could use the sppkg name from config/package-solution.json. This name must be unique because it's used for the name of .sppkg which is deployed to app catalog

SPFx already uses information from the project. By default, it sources the project version from package.json into web part version where in the manifest they use `*` to indicate that the version should be retrieved from package.json. I wonder if we could use a similar construct that is already familiar for folks who build on SPFx.

Reply:
Here's where I keep getting wrapped around the axle on this idea. None of those sources exist or are directly available within our library code.
We ship `mgt-element` which contains the `withDisambiguation` function. If I'm understanding the request here correctly it is to allow for developers to call `withDisambiguation()` as opposed to `withDisambiguation('foo')`. 
How can the `mgt-element` library code load `package.json` or `config/package-solution.json` from the calling application that it has been imported into?

AFAIK SPFx addresses this kind of challenge as it's part of the build toolchain and therefore can use known files in the calling application to configure the produced bundle.

Maybe there's something here that I'm totally missing.

Reply:
You're right. We'd need something like a webpack const that's replaced at build time, because none of the other files are available on runtime. Let me dig into the internal of SPFx to see if there's anything that we could use.

Reply:
We'd also need to have the build process explicitly work on a library in node_modules, which most build processes need to be configured to do as they use the contents of node_modules as is.
I think we need to differ this work until we can consult with the SPFx team to see if it's something that they want to take on.

Reply:
I checked with @AJIXuMuK and right now there's nothing in SPFx that we could use as a sane default. Let's revisit in the future should things change.

Reply:
Added #2022 to track this sane default concept. 

@sebastienlevert can you please take another pass at the docs we have now and see if we need to make any other adjustments to what we currently have here.